drivers/platform/stm32/stm32_spi.c

```diff
@@ -386,12 +424,134 @@ int32_t stm32_spi_write_and_read(struct no_os_spi_desc *desc,
+ */
+int32_t stm32_config_dma_and_start(struct no_os_spi_desc* desc,
+				   struct no_os_spi_msg* msgs,
+				   uint32_t len)
+{
+	struct stm32_spi_desc* sdesc = desc->extra;
+	struct no_os_dma_xfer_desc* rx_ch_xfer;
+	struct no_os_dma_xfer_desc* tx_ch_xfer;
+	struct no_os_dma_ch* tx_ch = sdesc->txdma_ch;
+	struct no_os_dma_ch* rx_ch = sdesc->rxdma_ch;
+	struct stm32_dma_channel* sdma_rx = rx_ch->extra;
+	struct stm32_dma_channel* sdma_tx = tx_ch->extra;
+	SPI_TypeDef* SPIx = sdesc->hspi.Instance;
+	int ret;
+	uint8_t i;
+
+	if (!desc || !msgs)
+		return -EINVAL;
+
+	rx_ch_xfer = no_os_calloc(len, sizeof(*rx_ch_xfer));
+	if (!rx_ch_xfer)
+		return -ENOMEM;
+
+	tx_ch_xfer = no_os_calloc(len, sizeof(*tx_ch_xfer));
+	if (!tx_ch_xfer) {
+		goto free_rx_ch_xfer;
+	}
+
+	for (i = 0; i < len; i++) {
+		tx_ch_xfer[i].src = msgs[i].tx_buff;
+		tx_ch_xfer[i].dst = &(SPIx->DR);
+		tx_ch_xfer[i].xfer_type = MEM_TO_DEV;
+		tx_ch_xfer[i].periph = NO_OS_DMA_IRQ;
+		tx_ch_xfer[i].length = msgs[i].bytes_number;
+
+		rx_ch_xfer[i].dst = msgs[i].rx_buff;
+		rx_ch_xfer[i].src = &(SPIx->DR);
+		rx_ch_xfer[i].periph = NO_OS_DMA_IRQ;
+		rx_ch_xfer[i].xfer_type = DEV_TO_MEM;
+		rx_ch_xfer[i].length = msgs[i].bytes_number;
+	}
+
+	rx_ch->id = sdma_rx->hdma;
+	rx_ch->sg_list = NULL;
+	if (sdma_rx->irq_num > 0)
+		rx_ch->irq_num = sdma_rx->irq_num;
+
+	ret = no_os_dma_config_xfer(sdesc->dma_desc, rx_ch_xfer, len, rx_ch);
+	if (ret)
+		goto remove_dma;
+
+	tx_ch->id = sdma_tx->hdma;
+	tx_ch->sg_list = NULL;
+	if (sdma_tx->irq_num > 0)
+		tx_ch->irq_num = sdma_tx->irq_num;
+
+	ret = no_os_dma_config_xfer(sdesc->dma_desc, tx_ch_xfer, len, tx_ch);
+	if (ret)
+		goto remove_dma;
+
+	ret = no_os_dma_xfer_start(sdesc->dma_desc, tx_ch);
+	if (ret)
+		goto abort_transfer;
+
+	ret = no_os_dma_xfer_start(sdesc->dma_desc, rx_ch);
+	if (ret)
+		goto abort_transfer;
+
+	if (tx_ch)
+		SET_BIT(sdesc->hspi.Instance->CR2, SPI_CR2_TXDMAEN);
+
+	if (rx_ch)
+		SET_BIT(sdesc->hspi.Instance->CR2, SPI_CR2_RXDMAEN);
+
+	if (sdesc->pwm_desc) {
+		ret = no_os_pwm_enable(sdesc->pwm_desc);
+		if (ret)
+			goto abort_transfer;
+	}
+
+	return 0;
+
+abort_transfer:
+	no_os_dma_xfer_abort(sdesc->dma_desc, tx_ch);
+	no_os_dma_xfer_abort(sdesc->dma_desc, rx_ch);
+remove_dma:
+	no_os_dma_remove(sdesc->dma_desc);
+free_tx_ch_xfer:
+	no_os_free(tx_ch_xfer);
+free_rx_ch_xfer:
+	no_os_free(rx_ch_xfer);
+
+	return ret;
+}
+
+/**
+ * @brief Configure and start a series of transfers using DMA. Wait for the
+ * 	  completion before returning.
+ * @param desc - The SPI descriptor.
+ * @param msgs - The messages array.
+ * @param len - Number of messages.
+ * @param callback - Function to be invoked once the transfers are done.
+ * @param ctx - User defined parameter for the callback function.
+ * @param is_async - Whether or not the function should wait for the completion.
+ * @return 0 in case of success, errno codes otherwise.
+ */
+int32_t stm32_spi_dma_transfer_sync(struct no_os_spi_desc* desc,
```

Comment:
What is the behavior of this transfer function? I see it doesn't wait for the transfer to finish before returning, but `callback` is ignored as well.

Reply:
I am trying to implement a non blocking function, the goal is to return from this API irrespective of whether the transfer/receive is complete or not. 
The callback is being registeredwithin the no_os_dma_config_xfer() and enabled via the no_os_dma_xfer_start()